File: js/arrays_strings/longest_substring_without_repeating_characters.js
// Use of sliding window technique

const longestSubstringWithoutDuplicateCharacter = (str) => {
    let l = 0;
    let r = 0;
    let ans = 0;
    let m = {};
    let n = str.length;
    while(l < n && r < n){
        let el = str[r];
        if(el in m){
            l = Math.max(l, m[el]+1);
        }
        m[el] = r;
        ans = Math.max(ans, r-l +1 ); 
        r++;
    }
    return ans;
}

let str = 'abcabcbb'; // 3
// let str = 'abba'; // 2
// let str = 'aaabbbc'; // 2
// let str = 'abcddabakjp'; // 5
console.log(longestSubstringWithoutDuplicateCharacter(str));

File: python/arrays_strings/longest_substring_without_repeating_characters.py
from typing import List

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        right = 0
        m = {}
        ans = 0
        n = len(s)
        while(left <n and right <n):
            el = s[right]
            if(el in m):
                left = max(left, m[el] +1)
            m[el] = right
            ans = max(ans, right-left +1)
            right+=1
        return ans





s = Solution()
# str = 'abcabcbb'; # 3
# str = 'abba'; # 2
# str = 'aaabbbc'; # 2
str = 'abcddabakjp'; # 5
ans = s.lengthOfLongestSubstring(str)
print(ans)